### PR TITLE
Improve validation of replicator job parameters

### LINKED ISSUE #4273: Port some replicator features from fdbmain
We had added a few nice features and fixed some bugs in fdbmain branch that may not be present in the main branch.

A few candidates are (checks are for the ones which have been ported to `main` already):

   - [ ] Remove duplicated VDU and job parsing code and replace with a single parse Erlang module. https://github.com/apache/couchdb/commit/b38d77fbada7cce7de288d2cdcca8839b09888f4
   - [x] Some proxy protocol handling: https://github.com/apache/couchdb/commit/cf71d00cb4b41da5527230d62b63e714c7777cf2 (ported in https://github.com/apache/couchdb/pull/3634)
   - [x] Bug combining custom headers in replicator jobs: https://github.com/apache/couchdb/commit/4246e9fb4d5e442944f43709e1cbec8f9ff02593  (https://github.com/apache/couchdb/commit/ce15da09a09407cefdf712eea4c2d20d3e83425f#diff-5e22b1fc94ae5fbbe09758f3efacc732f4bd89c82cddd5a26b0f6dfca0ec47e9)
   - [x] Fix bug in replicator when password or username contains `@` https://github.com/apache/couchdb/commit/b972f126c0266c88bc492ec13e7631a70fb5c1cd  (superseded by https://github.com/apache/couchdb/commit/1860ebbf2fa1731a62f3c9b107b2e52811489c1e)
   - [x] Jiffy encoding of atts_list maybe fail if it returns an io_list https://github.com/apache/couchdb/commit/bdfb129c1242e26cf312d2bc5cf1fe3af2e1e56d

### PR DESCRIPTION
**Pull request by nickva**
There are two main improvements:

  * Replace the auto-inserted replicator VDU with a BDU. Replicator already had
    a BDU to update the `"owner"` field, so plug right into it and validate
    everything we need there. This way, the validation and parsing logic is all
    in one module. The previously inserted VDU design doc will be deleted.

  * Allow constraining endpoint protocol types and socket options. Previously,
    users could create replications with any low level socket options. Some of
    those are dangerous and are possible "foot-guns". Restrict those options to
    a more usable set.

In addition to those improvements, increase test coverage a bit by explicitly
checking a few more parsing corner cases.

Fixes #4273

### CODE REVIEW

**Comment by nickva** [comment]
Note for the reviewer: the diff looks a bit larger than it is, but it's mostly because the parsing functions were moved to a separate parsing module - `couch_replicator_parse`. This idea was borrowed from the `fdbmain` replicator, and it seems like a better organization scheme.

**Review by jaydoane — APPROVED** [review]
+1 won't miss that VDU

File: src/couch_replicator/src/couch_replicator_parse.erl [hunk 1]
```diff
@@ -0,0 +1,750 @@
...
+        {checkpoint_interval, cfg_int("checkpoint_interval", 30000)},
+        {use_checkpoints,     cfg_boolean("use_checkpoints", true)},
+        {use_bulk_get,        cfg_boolean("use_bulk_get", true)},
+        {socket_options,      cfg_sock_opts()}
+    ].
+
+% Note: parse_rep_doc can handle filtered replications. During parsing of the
+% replication doc it will make possibly remote http requests to the source
+% database. If failure or parsing of filter docs fails, parse_doc throws a
+% {filter_fetch_error, Error} excation. This exception should be considered
```
[the remark this answers is not in the document]
Comment (nickva) [reply src/couch_replicator/src/couch_replicator_parse.erl:59]: Thanks!
